habitat_code/kobylarz_kmeans_review_code_20240313.py

```diff
@@ -0,0 +1,603 @@
+gv.tile_sources.EsriImagery() * gv.Polygons(huc_gdf).opts(
+    fill_alpha=0, line_color='blue', title='HUC 080902030506',
+    height=600, width=600
+)
+
+
+# I chose this watershed because it covers parts of New Orleans an is near the Mississippi Delta. Deltas are boundary areas between the land and the ocean, and tend to contain a rich variety of different land cover and land use types.
+# 
+# In the cell below, write a 2-3 sentence **site description** of this area that helps to put your analysis in context.
+
+# **Site Description**
+# 
+# HUC 080902030506 is a HUC 12 watershed located in the Mississippi Delta southeast of New Orleans. HUC 12 watersheds are the smallest geographical unit within the Watershed Boundary Dataset and are often analyzed to assess watersheds on a local level. This watershed consists of wetlands, open water, floodplains, upland islands, and some manmade features such as levees.
+
+# ## STEP 3: MULTISPECTRAL DATA
+
+# ### Search for data
+# 
+# **YOUR TASK:**
+#   1. Log in to the `earthaccess` service using your Earthdata credentials:
+#      ```python
+#      earthaccess.login(persist=True)
+#      ```
+#   2. Modify the following sample code to search for granules of the HLSL30 product overlapping the watershed boundary from May to October 2023 (there should be 76 granules):
+#      ```python
+#      results = earthaccess.search_data(
+#          short_name="...",
+#          cloud_hosted=True,
+#          bounding_box=tuple(gdf.total_bounds),
+#          temporal=("...", "..."),
+#      )
+#      ```
+
+# In[5]:
+
+
+# Search multispectral data from Earthdata
+
+earthaccess.login(persist=True)
+
+results = earthaccess.search_data(
+    short_name="HLSL30",
+    cloud_hosted=True,
+    bounding_box=tuple(huc_gdf.total_bounds),
+    temporal=("2023-05-01", "2023-09-30"),
+)
+
+
+# ### Compile information about each granule
+# 
+# I recommend building a GeoDataFrame, as this will allow you to plot the granules you are downloading and make sure they line up with your shapefile. You could also use a DataFrame, dictionary, or a custom object to store this information.
+# 
+# **YOUR TASK -- Write a function that:**
+#   1. For each search result:
+#       1. Get the following information (HINT: look at the ['umm'] values for each search result):
+#           - granule id (UR)
+#           - datetime
+#           - geometry (HINT: check out the shapely.geometry.Polygon class to convert points to a Polygon)
+#       2. Open the granule files. I recomment opening one granule at a time, e.g. with (`earthaccess.open([result]`).
+#       3. For each file (band), get the following information:
+#           - file handler returned from `earthaccess.open()`
+#           - tile id
+#           - band number
+#   2. Compile all the information you collected into a GeoDataFrame
+
+# In[5]:
+
+
+# Open the Earthaccess results so you only have to do it once
+@cached("open_results", False)
```

Comment:
I like how you are opening the results before caching -- lets you avoid some problems I was having with cached search results expiring.